gqcp/include/QCModel/HF/RHF.hpp

```diff
@@ -207,21 +213,22 @@ class RHF {
      *  @param sq_hamiltonian       the Hamiltonian expressed in an orthonormal basis
      *  @param N_P                  the number of electron pairs
      * 
-     *  @return the RHF orbital Hessian as a BlockRankFourTensor, i.e. an object with a suitable operator() implemented
+     *  @return the RHF orbital Hessian as a ImplicitRankFourTensorSlice, i.e. an object with a suitable operator() implemented
      */
-    static BlockRankFourTensor<Scalar> calculateOrbitalHessianTensor(const SQHamiltonian<Scalar>& sq_hamiltonian, const size_t N_P) {
+    static ImplicitRankFourTensorSlice<Scalar> calculateOrbitalHessianTensor(const SQHamiltonian<Scalar>& sq_hamiltonian, const size_t N_P) {
 
+        // Create an occupied-virtual orbital space.
         const auto K = sq_hamiltonian.dimension();
+        const auto orbital_space = OrbitalSpace::Implicit({{OccupationType::k_occupied, N_P}, {OccupationType::k_virtual, K - N_P}});  // N_P occupied spatial orbitals, K-N_P virtual spatial orbitals
 
-        BlockRankFourTensor<Scalar> hessian {N_P, K, 0, N_P,
-                                             N_P, K, 0, N_P};  // zero-initialize an object suitable for the representation of a virtual-occupied,virtual-occupied object (ai,bj)
+        auto hessian = orbital_space.template initializeRepresentableObjectFor<Scalar>(OccupationType::k_virtual, OccupationType::k_occupied, OccupationType::k_virtual, OccupationType::k_occupied);  // zero-initialize an object suitable for the representation of a virtual-occupied,virtual-occupied object (ai,bj)
 
         // Loop over all indices (ai,bj) to construct the orbital hessian
-        for (size_t a = N_P; a < K; a++) {
-            for (size_t i = 0; i < N_P; i++) {
+        for (const auto& a : orbital_space.indices(OccupationType::k_virtual)) {
+            for (const auto& i : orbital_space.indices(OccupationType::k_occupied)) {
 
-                for (size_t b = N_P; b < K; b++) {
-                    for (size_t j = 0; j < N_P; j++) {
+                for (const auto& b : orbital_space.indices(OccupationType::k_virtual)) {
+                    for (const auto& j : orbital_space.indices(OccupationType::k_occupied)) {
```

Comment:
@dariatols @xdvriend @guacke This is the API change that I'm most proud of. I think this is a really nice example of OOP doing its job with encapsulation, enhancing readability and making implementations less error-prone.

Reply:
Awesome! Less error-prone and enhances the readability of the code!

Reply:
Awesome! Now you can worry less about which indices are occupied/virtual. Nice!